pkg/pipelines/webhook/webhook_test.go

```diff
@@ -127,3 +129,46 @@ func TestGetGitRepoURL(t *testing.T) {
 		})
 	}
 }
+
+func TestGetAccessToken(t *testing.T) {
+	keyring.MockInit()
+	optionTests := []struct {
+		name             string
+		gitRepo          string
+		envVarPresent    bool
+		tokenRingPresent bool
+		hostName         string
+		expectedToken    string
+	}{
+		{"Github Token not present in keyring/env-var", "https://github.com/example/test.git", false, false, "github.com", ""},
+		{"Github Token present in env-var", "https://github.com/example/test.git", true, false, "github.com", "abc123"},
+		{"Github Token present in keyring", "https://github.com/example/test.git", false, true, "github.com", "xyz123"},
+		{"Github Token defaults to keyring although env-var present", "https://github.com/example/test.git", true, true, "github.com", "xyz123"},
+		{"Gitlab Token not present in keyring/env-var", "https://gitlab.com/example/test.git", false, false, "gitlab.com", ""},
+		{"Gitlab Token present in env-var", "https://gitlab.com/example/test.git", true, false, "gitlab.com", "abc123"},
+		{"Gitlab Token present in keyring", "https://gitlab.com/example/test.git", false, true, "gitlab.com", "xyz123"},
+		{"Gitlab Token defaults to keyring although env-var present", "https://gitlab.com/example/test.git", true, true, "gitlab.com", "xyz123"},
+	}
+
+	for _, tt := range optionTests {
+		if tt.envVarPresent {
+			err := os.Setenv("TESTGITTOKEN", "abc123")
+			if err != nil {
+				t.Errorf("Error in setting the environment variable")
+			}
+			defer os.Unsetenv("TESTGITTOKEN")
```

Comment:
You're not deferring in the right bit, you should be deferring right after you set it.

But, in this case, it will work because t.Errorf doesn't cause the immediate exit of the function.

But, you still have it being unset below, despite deferring it being cleared.

Can I suggest a small bit of rework here, execute each of the tests in a t.Run(...) call, which means that the defer will execute when that completes, otherwise you're deferring the token being reset a lot.

If this doesn't appeal, then you can defer the os.Unsetenv("TESTGITTOKEN") at the start of the function, that way when it exits, it will get unset.

Reply:
Thanks @bigkevmcd , this change worked.